### ISSUE TITLE
[2018] Rate Distortion For Model Compression: From Theory To Practice

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#36]
## 論文リンク
https://arxiv.org/abs/1810.06401

## 公開日（yyyy/mm/dd）
2018/10/09

## 概要
model compression に rate distortion theory を適用するもの。
詳しくはあとで書く。

**Comment by yoheikikuta** [comment]
最近 model compression の話を見てないなとか思いながら ICML 2019 accepted papers を眺めていたら発見した。
information bottleneck を読んでいた時に rate distortion theory を勉強していたりしたので、予備知識としてもちょうど良さそうだし。

情報量の観点から weight の pruning や quantization の効率的なやり方を決定する、というのも方向性として興味深いところである。

**Comment by yoheikikuta** [comment]
モチベーションはシンプルで、どうすれば効率的に model compession が実現できるかを理論的根拠に基づいて明らかにしたいというもの。

pruning も quantization も既に存在しているが、典型的なものは特に理論的根拠はない。
前者は weight の絶対値の小さいものを落としたり feature map の誤差を小さく保ちつつ channel を落としたりする。
後者は単純に float32 を int8 にしたりする。

model compression は言ってみれば、model weight parameter を表現する bit 数を少なくしつつ compress 前の元モデルからの distortion を小さくする、というものであるとも言えるのでまさに rate distortion theory を適用できそうな問題設定である。

**Comment by yoheikikuta** [comment]
まずは rate distortion theorem とはどういうものだったかの復習。

以下のように encoder で codeword へ encode して、それを decode して元の空間に戻すような状況を考える。

<p align="center">
  <img src="https://i.imgur.com/B70OIzh.png">
</p>

このとき、data compression に対する rate distortion theorem は以下のようにかける。
d は distortion function で I は相互情報量である。

<p align="center">
  <img src="https://i.imgur.com/Ucmes5v.png">
</p>

相互情報量を小さくする、すなわち 2 つの確率変数が互いに依存する度合いを小さくするということで compression を表現している。
何の制限もなければ例えば一点に map して互いを独立にすれば相互情報量をゼロにできるが、それでは意味がなくなってしまう。
そこで distortion のある上限値 D を定め、それを満たすような制限を加えることで元の情報を良い感じに残しつつ compress したいというものになっている。

**Comment by yoheikikuta** [comment]
具体例として $` X ~ N(0, σ^2) `$ の場合を考える。distortion function は square distrotion とする。
この場合 $` R(D) `$ は以下のようになる。

<p align="center">
  <img src="https://i.imgur.com/WVIVEqM.png">
</p>

$` D > σ^2 `$ の場合は元の gaussian の分散よりも distortion が大きくなって良いので、$` \hat{X} = 0 `$ という一点 mapping を考えれば条件を満たせてしまうので trivial である。
一方で、$` D \leq σ^2 `$ の場合にこの結果になるのは以下のように証明できる。

<p align="center">
  <img src="https://i.imgur.com/jUy8bBx.png">
</p>

これはだいぶ適当にやってしまったが、離散でなく連続でも同様である。
$` \hat{X} `$ と $` X - \hat{X} `$ を独立にとるようにしているところがややテクニカルだが、結果は直感的には難しくなく、gaussian の entropy を扱うことになるので D を大きくするにしたがって log 的に $` R(D) `$ が小さくなっていくという振る舞いとなる。

**Comment by yoheikikuta** [comment]
復習はこれくらいにしてこの手法を model compression に適用することを考える。
compress するのは weight であるので、その weight で parametrize されたモデル $` f_w `$ を使って distortion function をうまく定義すればよいということになる。
これは形式的に書けば以下のようになるだろう。

- regression の場合

<p align="center">
  <img src="https://i.imgur.com/ucOWLZV.png">
</p>

- classification の場合

<p align="center">
  <img src="https://i.imgur.com/aWGvXq1.png">
</p>

この f_w はどこの出力なのかというのはやや抽象的だが、unsupervised であればどこかの feature map で、supervised であれば loss function にする、という感じになる。

これらの distortion function を適切に使い分けるという前提で、model compression のための rate distortion theorem は以下のように書くことができる。

<p align="center">
  <img src="https://i.imgur.com/wcNaBuq.png">
</p>

**Comment by yoheikikuta** [comment]
ここまでは抽象的な議論なので特に問題はないが、 具体的にどのように計算するのかということに踏み込もうとすると、議論の進め方に注意が必要である。

distortion function を定めて相互情報量をガッと計算できればそれで問題なのだが、複雑な NN でそれを実行するのは困難である。

そこで、この論文では以下のように議論を展開する。

- linear regression の場合でさらに weight を gaussian からサンプルしたという仮定の下で、$` R(D) `$ の下限を解析的に導出する
- その結果を基に weight が満たすべき条件を抽出する
- 複雑な（つまり non-linear な）ケースにも一次のテイラー展開を介して適用して linear regression の場合の条件に類するものを作る
- （一層の ReLU NN では求めた条件が確かに optimal であることを証明する）
- 実際の NN に適用して実験してみて、従来手法よりも効果的に model compression が実現できることを示す

なので、一般の場合にがっちり理論武装された方法ではないのだが、シンプルな場合に理論的に保証されたものを少し拡張して適用してみて、それは実験的にこれまでの結果よりも良いものである、という流れである。

**Comment by yoheikikuta** [comment]
ということで linear model: $` f_w = w^T x `$ の場合から考える。
仮定として、データ $` x ∈ R^m `$ は平均 0 で、分散共分散行列は diagonal とする $` E_X[X_i^2] = λ_{x,i} `$ で、weight $` W ∈ R^m `$ は $` N(0, \sum_W) `$ からサンプルしたものとする。

この時、以下の定理が成り立つ。

<p align="center">
  <img src="https://i.imgur.com/BXM1Ofa.png">
</p>

これの証明は Appendix A を真面目に読めば分かる。結構丁寧に書いてあるのでここでは割愛するが、重要なのは条件付きエントロピーは条件を増やしてもエントロピーが増加することはない（条件を増やすことは不確かさを減らす方向のみに働く）ので下限をつけれるという点である。

これの解釈というか考え方として weighted water-filling があるのでそれだけ見ておこう。
簡単のため、次元が 3 の場合で  $` \sum_W = \sum_X = diag[3, 2, 1] `$ という場合を考える。
rate-distortion function の下限は一項目は const なので二項目が D の変更とともにどのように変化するのかが焦点となる。
以下の図のような box を考えて、ある $` D `$ を決めた時に各 box に同体積の水を注いでいくということを考える。水位が D_i であり、これの log の和が大きくなるほど R(D) が小さくできるということになっている。小さい箱は効率的に水位が高められるので最初のうちは R(D) の減少具合が鋭いが、小さい箱が埋まると効率が悪くなっていくので R(D) の減少具合が鈍くなっていくという具合である。

<p align="center">
  <img src="https://i.imgur.com/5glp1A9.png">
</p>

そしてこの下限が実際に達成可能である、ということも証明されている。
具体的なアルゴリズムは以下。ここは詳細は置いといてちゃんと下限が達成可能なアルゴリズムがあるということだけ押さえておく。

<p align="center">
  <img src="https://i.imgur.com/k0nNtN4.png">
</p>

---
やはり以降の話をするときに証明で使われている不等式が理解できていないといけないので、論文の appendix のものをそのまま貼っておく。
以下の補題は rate-distortion function に出てくる相互情報量を下から bound するためのものである。使っているのは条件付きエントロピーの性質と、differential entropy が gaussian で maximize されるというものである。

<p align="center">
  <img src="https://i.imgur.com/qe5O8Nw.png">
</p>

証明の本筋はこの補題を使ってdistortion function の期待値が $` D `$ 以下という制限下で具体的にどういう形で $` R(D) `$ を表現できるかを示すものであり、これもまるっとコピーしてくると以下。

<p align="center">
  <img src="https://i.imgur.com/4AGkQCl.png">
</p>

うーむ、丸ごとコピーしまくってるのでめちゃ長くなってしまうな。
以降の話と特に関連が深いのは $` D_i `$ が登場するところで、全体の distorition level $` D `$ が与えられていて、その下限を捉えているものが $` \sum λ_{x,i} D_i `$ になっているという点である。

$` R(D) `$ を求めるというのは、distortion $` D `$ が制限として与えられた上で rate $` R `$ を最小化する問題と捉える、という見方ではなくて逆に rate $` R `$ が制限として与えられた上で distortion $` D `$ を最小化する問題と捉える、という見方もできる。

つまり、$` D_i `$ を求める作業を標語的に言えば、rate $` R `$ が制限として与えられた上で distortion function の期待値を minimize するということができる。

**Comment by yoheikikuta** [comment]
linear model の議論から重要な点を抽出しよう。
R(D) の下限が実現したい理想的な状態であるので、下限を満足するときにどういう条件になっているか、という点である。そしてそれは以下のようになる。

<p align="center">
  <img src="https://i.imgur.com/pO9WbsS.png">
</p>

特に独立の方の条件は $` E_{W, \hat{W}} [WˆT Σ_X (W - \hat{W})] `$ と書ける。
条件をまとめて書いてしまってやや分かりづらいが、こいつを複雑な NN の場合でも成り立っていてほしいルールとする。

さらに、

> つまり、$` D_i `$ を求める作業を標語的に言えば、rate $` R `$ が制限として与えられた上で distortion function の期待値を minimize するということができる。

であったので、これもルールとして採用しよう。

以上二つをまとめると以下のように論文でいうところの golden rule が定まる。

<p align="center">
  <img src="https://i.imgur.com/oWZPN9n.png">
</p>

これらは分散共分散行列が含まれているので linear model での rule である。
以降では、これを non-linear model に適用していくということを考える。

**Comment by yoheikikuta** [comment]
これまでは $` W `$ は確率変数として取り扱っていたが、実際には pre-trained model が一個あってそこの weight を prune したり quantize したりするだけなので、こいつらは何か一個の実数値 w として書こう。

regression の場合（つまり distortion function が feature map の二乗誤差）から考える。distortion function において $` f_w `$ と compress 後の $` f_{\hat{w}} `$ は non-linear でそのままでは扱い難いので、後者を $` w `$ の周りで Taylor 展開して以下のように書こう。

<p align="center">
  <img src="https://i.imgur.com/7dBmqvz.png">
</p>

次に classification の場合。これはちょろっと計算すると以下のように weight important matrix が得られる。

<p align="center">
  <img src="https://i.imgur.com/rYhDcsc.png">
</p>

以上を踏まえ、non-linear model の場合の rule を以下のように準備する。

<p align="center">
  <img src="https://i.imgur.com/w2UK4II.png">
</p>

あとはこれを 

- pruning ならば pruning する割合が与えられている条件
- quantization ならば centroid の数が与えられているという条件

で $`\hat{w}`$ を求めていけばよい（そういうものが rule 1 を満たしていることも示せるが、まあその事実だけ知っとけばいいかなということで割愛）。

気をつけなければならない点は、このような non-linear model での rule はあくまで linear model の結果を理論的な根拠なく拡張したもので、optimal な $` R(D) `$ であるという保証は無いという点である。

この論文では、 1-layer RELU ならば optimal になっているということを示してこの rule が良さそうだという雰囲気を出しておいて、あとは比較的複雑なモデルを実験で試して確かに有用だと示す、という流れになる。
前者はやや手間なので割愛し、実験結果を見ていくとする。

**Comment by yoheikikuta** [comment]
実験としては以下。

- 3-layer fully connected NN on MNIST
- CNN with 5 convolutional layers and 3 fully connected layers on CIFAR10 and CIFAR100 (CIFAR100 は appendix に載っていて結果の傾向は変わらないのでここでは割愛しとく)

さらにこれらを unsupervised の場合と supervised （ラベルが使えるので loss function を使える）の場合それぞれで試す。

accuracy-compression rate や cross entropy-compression rate の plane を見ていく。

**Comment by yoheikikuta** [comment]
**unsupervised の場合**

式を書けないのが不便なのでサボってコピペすると、以下のようなセットアップになる。
簡単のため off-diagonal は無視し、その結果従来手法に対して weight ごとに異なる importance が与えられたようなものになっている。quantization も importance 付きの k-means で centroid を決める。

<p align="center">
  <img src="https://i.imgur.com/Av42Cyw.png">
</p>

- pruning の結果
MNIST の accuracy に関しては従来手法に負けてるという結果だが、まあ MNIST だしなという感じ。ちなみに横軸は差が見やすいようにそれぞれのケースでバラバラの範囲なので注意。compress を激しくした場合に従来手法よりも特に優れた結果になっている。

<p align="center">
  <img src="https://i.imgur.com/t1DSgH9.png">
</p>

- quantization の結果
傾向としては pruning と同じような感じ。compression ratio は雑に言うとビット換算でどれくらい compress されるかで測っている。

<p align="center">
  <img src="https://i.imgur.com/0dzy3Md.png">
</p>

**Comment by yoheikikuta** [comment]
**supervised の場合**

この場合 $` f_w `$ として loss function を使うことができる。
pruning の場合と同様に off-diagonal を落とせば、importance が loss function の微分の二乗で与えられることになる。$` E `$ は $` E_{X, Y} `$ の意。

<p align="center">
  <img src="https://i.imgur.com/f5Qide0.png">
</p>

直接的に loss function を使えるので結果がよくなりそうな気がするが、ここで気になるのは loss function の微分を使うってことはよく学習されてるモデルならゼロに近くなってうまく work しないのでは？ということである。
（実際にはこれだけでも結構うまくいくのだが）処方箋としては Taylor 展開の二次まで取って Hessian も使うことができる。ということで以下のパターンをそれぞれ試す。

<p align="center">
  <img src="https://i.imgur.com/o6eYFjJ.png">
</p>

このように拡張した場合、pruning や quantization の方法も変更が必要。
pruning は項が増えても単にそれを考慮した上で値が小さい w_i から切っていけばいいが、quantization は四次の項が入ってくるとどうやって k-means をするか真面目に考える必要があり、それは appendix に載ってる。ぱっと見で概念的に難しいことはなさそうなので、そこはちゃんと追ってない。

- pruning の結果
unsupervised よりも良さそうな結果。gradient + hessian は MNIST は更に良くなりそうだけど CIFAR10 はそんなに変わらなそうかなという感じ。

<p align="center">
  <img src="https://i.imgur.com/JyXOkL8.png">
</p>

- quantization の結果
これは gradient + hessian がかなり良い結果を示している。これくらい improve があるならなかなか素晴らしいと思うが、hessian 計算するのは時間が掛かると言って appendix の CIFAR100 では gradient + hessian の結果がないのが残念。どれくらい一般的かを知りたいところなのに。

<p align="center">
  <img src="https://i.imgur.com/91H8dGZ.png">
</p>

**Comment by yoheikikuta** [comment]
ということで一通り読んでみた。
rate distortion theory をある程度勉強したことがあるので自分に取っては読みやすい論文だった。
理論に基づいて進めるけど、一般の場合は難しいので簡単な場合からヒントを得てそれを拡張してみたら確かに有用でした、という話だった。

model compression でこういう話が進んで真面目に効率性が議論できるようになっていくのはなかなか良いと思う。
